File: src/hey/StringtoInteger.java
/*
    problem: String to Integer (atoi)
    author: everpuck
    date: 2/16/2019
*/


class Solution {
    public int myAtoi(String str) {
        String trimStr = str.trim();
        int sign = 1;
        double ret = 0;

        for (int i = 0; i < trimStr.length(); i++) {
            if (i == 0) {
                if (trimStr.charAt(i) == '-') {
                   sign = -1;
                }
                else if (trimStr.charAt(i) == '+') {
                    continue;
                }
                else if (trimStr.charAt(i) > 47 && trimStr.charAt(i) < 58) {
                    ret = trimStr.charAt(i) - 48;
                }
                else {
                    return 0;
                }
            }
            else {
                if (trimStr.charAt(i) > 47 && trimStr.charAt(i) < 58) {
                    ret = ret * 10 + trimStr.charAt(i) - 48;
                }
                else {
                    break;
                }
            }
        }
        ret *= sign;
        if (ret < Integer.MIN_VALUE) {
            ret = Integer.MIN_VALUE;
        }
        if (ret > Integer.MAX_VALUE) {
            ret = Integer.MAX_VALUE;
        }
        return (int)ret;
    }
}


public class StringtoInteger {
    public static void main(String[] args) {
        // String str = " +-0123abc0";
        // String str = " -01239abc0";
        // String str = "      +123abc0";
        // String str = " ";
        // String str = "-2147483649";
        String str = "9223372036854775808";
        Solution solution = new Solution();
        int ret = solution.myAtoi(str);
        System.out.println(ret);
    }
}

File: src/hey/StringtoInteger.py
# problem: String to Integer (atoi)
# author: everpuck
# date: 2/15/2019


class Solution:
    def myAtoi(self, str: 'str') -> 'int':
        ret_str = str.lstrip(' ')
        ret = 0
        sign = 1
        for i, c in enumerate(ret_str):
            if i == 0:
                if c == "-":
                    sign = -1
                elif c == "+":
                    continue
                elif c.isdigit():
                    ret = int(c)
                else:
                    return 0
            else:
                if c.isdigit():
                    ret = ret*10 + int(c)
                else:
                    break
        ret *= sign
        min_value = -2**31
        max_value = 2**31 - 1
        if ret < min_value:
            return min_value
        elif ret > max_value:
            return max_value
        return ret


if __name__ == '__main__':
    _str = 'words and 987'
    _str = '4193 with words'
    _str = '    -042'
    _str = ''
    solution = Solution()
    print(solution.myAtoi(_str))

